instance/models/server.py

```diff
@@ -76,54 +143,52 @@ def exclude_terminated(self):
         """
         Filter out terminated servers from the queryset
         """
-        return self.filter(~Q(status=Server.TERMINATED))
+        return self.filter(~Q(_status=Status.Terminated.state_id))
 
 
 class Server(ValidateModelMixin, TimeStampedModel):
     """
     A single server VM
     """
-    NEW = 'new'
-    STARTED = 'started'
-    ACTIVE = 'active'
-    BOOTED = 'booted'
-    PROVISIONING = 'provisioning'
-    REBOOTING = 'rebooting'
-    READY = 'ready'
-    LIVE = 'live'
-    STOPPING = 'stopping'
-    STOPPED = 'stopped'
-    TERMINATING = 'terminating'
-    TERMINATED = 'terminated'
-
-    STATUS_CHOICES = (
-        (NEW, 'New - Not yet loaded'),
-        (STARTED, 'Started - Running but not active yet'),
-        (ACTIVE, 'Active - Running but not booted yet'),
-        (BOOTED, 'Booted - Booted but not ready to be added to the application'),
-        (PROVISIONING, 'Provisioning - Provisioning is in progress'),
-        (REBOOTING, 'Rebooting - Reboot in progress, to apply changes from provisioning'),
-        (READY, 'Ready - Rebooted and ready to add to the application'),
-        (LIVE, 'Live - Is actively used in the application and/or accessed by users'),
-        (STOPPING, 'Stopping - Stopping temporarily'),
-        (STOPPED, 'Stopped - Stopped temporarily'),
-        (TERMINATING, 'Terminating - Stopping forever'),
-        (TERMINATED, 'Terminated - Stopped forever'),
+    Status = Status
+    status = ResourceStateDescriptor(state_classes=Status.states, default_state=Status.New, model_field_name='_status')
+    _status = models.CharField(
+        max_length=20,
+        default=status.default_state_class.state_id,
+        choices=status.model_field_choices,
+        db_index=True,
+        db_column='status',
     )
+    # State transitions:
+    _status_to_started = status.transition(from_states=Status.New, to_state=Status.Started)
+    _status_to_active = status.transition(from_states=Status.Started, to_state=Status.Active)
+    _status_to_booted = status.transition(from_states=Status.Active, to_state=Status.Booted)
+    _status_to_provisioning = status.transition(from_states=Status.Booted, to_state=Status.Provisioning)
+    _status_to_rebooting = status.transition(
+        from_states=(Status.Active, Status.Booted, Status.Ready, Status.Provisioning), to_state=Status.Rebooting
+    )
+    _status_to_ready = status.transition(
+        from_states=(Status.Active, Status.Booted, Status.Provisioning, Status.Rebooting), to_state=Status.Ready
+    )
+    _status_to_terminated = status.transition(to_state=Status.Terminated)
```

Comment:
@e-kolpakov I agree. The problem is that as written, `server.state._state_manager` doesn't have access to the `resource` instance, so couldn't hold transition methods. But since I wanted to implement `ResourceStateDescriptor` as a descriptor, I didn't yet see a simple way to organize the transitions nicely other than what I already put in this PR. Open to suggestions here.